### ISSUE TITLE
Incorrect use of _Inout_ ?

### ISSUE THREAD

**Issue by scottnoone** [MicrosoftDocs/cpp-docs#4487]
Hi,

The first corrected example says that we should use _Inout_ instead of _In_:

void Func2(_Inout_ PCHAR p1)
{
    if (p1 == NULL)
        return;

    *p1 = 1;
}

My understanding is that the _Inout_ here means the function expects *p1 to be valid on input and will write the contents of *p1 on output. So, based on that annotation, the following code generates a CA warning due to passing an uninitialized variable as an input:

    char foo;
    Func2(&foo);

But that's not really the intent of that function. It just wants p to be non-NULL and will write the content on success. Shouldn't the correct annotation then be _Out_opt_?

void Func2(_Out_opt_ PCHAR p1)
{
    if (p1 == NULL)
        return;

    *p1 = 1;
}

Thanks,

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: a052b3f2-97ef-3c75-3b0c-432a072eb76b
* Version Independent ID: 1581c417-5ef9-e357-3e29-0b8bd453ad54
* Content: [Best practices and examples (SAL)](https://learn.microsoft.com/en-us/cpp/code-quality/best-practices-and-examples-sal?view=msvc-170)
* Content Source: [docs/code-quality/best-practices-and-examples-sal.md](https://github.com/MicrosoftDocs/cpp-docs/blob/main/docs/code-quality/best-practices-and-examples-sal.md)
* Product: **visual-cpp**
* Technology: **cpp-tools**
* GitHub Login: @TylerMSFT
* Microsoft Alias: **twhitney**

**Comment by TylerMSFT** [comment]
Hi @scottnone,

Thanks for catching this. 

I think you are right. The corrected version will be live tomorrow using `_Out_`. `  The reason I didn't go with your first suggestion of `_Out_opt_` is probably the same reason you changed your mind. A NULL pointer isn't allowed, it is merely tolerated by defensively coding around the possibility.

**Comment by TylerMSFT** [comment]
@scotnoone, yeah, it is a gray area and I went back and forth as well. I think you can make the case both ways. But I decided to hang my hat on this statement: 

If the caller isn't allowed to pass in a null pointer, use `_In_`  or `_Out_` instead of `_In_opt_` or `_Out_opt_`. This applies even to a function that checks its parameters and returns an error if it is NULL when it shouldn't be. Although having a function check its parameter for unexpected NULL and return gracefully is a good defensive coding practice, it doesn't mean that the parameter annotation can be of an optional type (_*Xxx*_opt_).

In the example, the function doesn't crash because of the NULL guard. But you can say that the function doesn't fulfill its purpose if you send NULL because doesn't succeed in setting p1 to anything. In that sense, it doesn't tolerate NULL, so I decided `_Out_` is the way to go.

What are your thought about that?

**Comment by TylerMSFT** [comment]
@scottnoone I am persuaded by your argument. I've also updated the doc to reflect some of this thinking in the example:

```cpp
// Correct
// _Out_opt_ because the function tolerates NULL as a valid argument, i.e.
// no error is returned. If the function didn't check p1 for NULL, then
// _Out_ would be the better choice
void Func2(_Out_opt_ PCHAR p1)
{
    if (p1 == NULL)
        return;

    *p1 = 1;
}
```

For people wondering about this issue we are discussing, hopefully the doc is a little clearer by a) not being wrong ;-) and b) calling out the subtlety of `_Out_` vs `_Out_opt_` .

Thanks for engaging on this.

**Comment by scottnoone** [comment]
Awesome, looks good to me!

SAL is such a great tool but it’s daunting for people. Having clear examples demonstrating the intended usage of the annotations is great so I’m glad to help.

Thanks!

-scott

—
From mobile

On Mar 30, 2023, at 5:43 PM, Tyler Whitney ***@***.***> wrote:

﻿

@scottnoone<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fscottnoone&data=05%7C01%7Csnoone%40osr.com%7C33e5a99f94904b1e845b08db3167ce38%7C033b9c3463d541cab30b3edc5d617996%7C0%7C0%7C638158094153814276%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=gDRMaPpMB2PJ4oupHHXUK%2FzSF1eBikVS%2F9NgTFJxfuc%3D&reserved=0> I am persuaded by your argument. I've also updated the doc to reflect some of this thinking in the example:

// Correct
// _Out_opt_ because the function tolerates NULL as a valid argument, i.e.
// no error is returned. If the function didn't check p1 for NULL, then
// _Out_ would be the better choice
void Func2(_Out_opt_ PCHAR p1)
{
    if (p1 == NULL)
        return;

    *p1 = 1;
}

For people wondering about this issue we are discussing, hopefully the doc is a little clearer by a) not being wrong ;-) and b) calling out the subtlety of _Out_ vs _Out_opt_ .

Thanks for engaging on this.
